Add ability to convert solvability info to descriptor

sipa opened the pull request:
This PR adds functionality to convert a script to a descriptor, given a `SigningProvider` with the relevant information about public keys and redeemscripts/witnessscripts.

The feature is exposed in `listunspent`, `getaddressinfo`, and `scantxoutset` whenever these calls are applied to solvable outputs/addresses.

This is not very useful on its own, though when we add RPCs to import descriptors, or sign PSBTs using descriptors, these strings become a compact and standalone way of conveying everything necessary to sign an output (excluding private keys).

Unit tests and rudimentary RPC tests are included (more relevant tests can be added once RPCs support descriptors).

Fixes #14503.

sipa committed:
Add support for inferring descriptors from scripts

sipa committed:
Add Descriptor::IsSolvable() to distinguish addr/raw from others

sipa committed:
Add tests for InferDescriptor and Descriptor::IsSolvable

sipa committed:
Add descriptors to listunspent and getaddressinfo + tests

ryanofsky commented on src/script/descriptor.cpp:
```diff
@@ -298,6 +303,7 @@ class ConvertorDescriptor : public Descriptor
     ConvertorDescriptor(std::unique_ptr<Descriptor> descriptor, const std::function<CScript(const CScript&)>& fn, const std::string& name) : m_convert_fn(fn), m_fn_name(name), m_descriptor(std::move(descriptor)) {}
 
     bool IsRange() const override { return m_descriptor->IsRange(); }
+    bool IsSolvable() const override { return true; }
```
Should this be returning m_descriptor->IsSolvable() instead of true? Maybe add an explanatory comment if returning true here is actually correct.

ryanofsky commented on src/script/descriptor.h:
```diff
@@ -32,6 +32,9 @@ struct Descriptor {
     /** Whether the expansion of this descriptor depends on the position. */
     virtual bool IsRange() const = 0;
 
+    /** Whether this descriptor has all information about signing (ignoring private keys). */
```
Would it be useful to mention that raw and address descriptors are always considered unsolvable? It seems like this would make the description more concrete, and make it clear that a raw P2PK descriptor is not considered solvable (for example).

ryanofsky commented on src/script/descriptor.h:
```diff
@@ -32,6 +32,9 @@ struct Descriptor {
     virtual bool IsRange() const = 0;
 
+    /** Whether this descriptor has all information about signing (ignoring private keys). */
+    virtual bool IsSolvable() const = 0;
```
Curious where this might be used in the future, since right now it is only called by test code.

sipa replied:
For example an `descriptorprocesspsbt` RPC can complain if a non-solvable descriptor is passed (which is useless for its purpose).

ryanofsky commented on src/test/descriptor_tests.cpp:
```diff
@@ -99,8 +99,17 @@ void Check(const std::string& prv, const std::string& pub, int flags, const std:
 
+                /* Infer a descriptor from the generated script, and verify its solvability and that it roundtrips. */
+                auto inferred = InferDescriptor(spks[n], script_provider);
+                BOOST_CHECK_EQUAL(!(flags & UNSOLVABLE), inferred->IsSolvable());
```
This is a very picky comment, but maybe reverse the arguments so the test more consistently uses the CHECK_EQUAL(actual, expected) form.

sipa replied:
Done.

ryanofsky reviewed (approved):
utACK 002bdef765a7cdb4dcf66fa186a4c4a1a7fc7b71

sipa commented:
Addressed all nits, and added some comments.

Sjors commented:
Concept ACK. Lightly tested a5e0b62; e.g. `getaddressinfo` shows a descriptor now.

The descriptors only show the public key, which is consistent with `getaddressinfo` not showing private keys and seeds. Maybe add a `descriptor` boolean argument to `dumpprivkey` as a way to obtain descriptors with a private key?

promag commented on test/functional/wallet_address_types.py:
```diff
@@ -252,6 +261,14 @@ def run_test(self):
+                        assert_equal(utxo['descriptor'], descriptors[to_node][1])
+                        found = True
+                        break
+                assert(found)
```
`assert found`

sipa replied:
Fixed.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -21,6 +21,7 @@
 #include <rpc/server.h>
 #include <rpc/util.h>
 #include <script/sign.h>
+#include <script/descriptor.h>
```
nit sort.

sipa replied:
Fixed.

promag reviewed (commented):
For large wallets this will make `listunspent` perform a lot worse?

ryanofsky commented:
utACK a5e0b620062bf00a5f24d195bcaf52839681cb5d. Only changes since last review were making some suggested changes and expanding the InferDescriptor.

It would also be nice to add release notes for the RPC changes in this PR.

> For large wallets this will make listunspent perform a lot worse?

It seems like the new work listunspent is doing is similar to the work it was doing before with calling `ExtractDestination` and `pwallet->GetCScript`.

sipa committed:
Add release notes

sipa commented:
Sorry for the big change, but I realized that after #14150, we'll want to integrate origin information into the inferred descriptors this PR brings, so I went ahead and did that (it's based on 14150 now).

sipa committed:
Add matching descriptors to scantxoutset output + tests

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3539,6 +3545,8 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
             "  \"scriptPubKey\" : \"hex\",       (string) The hex-encoded scriptPubKey generated by the address\n"
             "  \"ismine\" : true|false,        (boolean) If the address is yours or not\n"
             "  \"iswatchonly\" : true|false,   (boolean) If the address is watchonly\n"
+            "  \"issolvable\" : true|false,    (boolean) Whether we know how to spend coins sent to this address, ignoring the possible lack of private keys\n"
```
I think it would be nicer to call this "solvable" instead of "issolvable" for consistency with `listunspent`. This would be less consistent with other fields here like "iswatchonly", but I think it's worse to have different names for the same field than different styles of names for different fields.

sipa replied:
Done.

ryanofsky commented on test/functional/wallet_address_types.py:
```diff
@@ -95,34 +96,42 @@ def get_balances(self, confirmed=True):
             assert(not info['isscript'])
             assert(not info['iswitness'])
             assert('pubkey' in info)
+            assert_equal(self.remove_origin(info['descriptor']), "pkh(%s)" % info['pubkey'])
```
It seems like it'd be good to check inferred origins instead of skipping them. If it isn't easy to do this, maybe you could add a comment to the test saying why.

sipa replied:
Done, rewritten the test to compare with information put in PSBT (which also reports fingerprint and path in `decodepsbt`).

ryanofsky reviewed (commented):
utACK cf4b935a0301972d0e1fc6a5cb32c6ff564dbbbc. Changes: rebasing over #14150, adding `InferPubkey()` and a test for the inferred origin field, making the address_types test ignore origin information in inferred descriptors, and adding release notes.

Sjors commented:
Lightly tested e6e4704; `getaddressinfo` descriptor now includes origin info.

ryanofsky reviewed (approved):
utACK e6e47041a70d48f53dd2016cd5ae1cd822c5d3c6. Changes since last review are rebase, `'solvable`' naming, and more comprehensive test code that compares `getaddressinfo` descriptors with `listunspent` descriptors and PBST `bip32_derivs` paths.

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/14477/commits/e6e47041a70d48f53dd2016cd5ae1cd822c5d3c6

sipa commented:
Rebased, renamed the descriptor output field to `desc` (for consistency with the name of the input argument to `scantxoutset`), and added the output to `scantxoutset` as well (fixing #14503).

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -2140,7 +2142,11 @@ UniValue scantxoutset(const JSONRPCRequest& request)
                 }
-                needles.insert(scripts.begin(), scripts.end());
+                for (const auto script : scripts) {
+                    std::string desc = InferDescriptor(script, provider)->ToString();
```
In commit "Add matching descriptors to scantxoutset output + tests" (fe9b68a81a6f3a9abac7b3b7740d5019129d4fff)

I think I agree with @Sjors in https://github.com/bitcoin/bitcoin/pull/14477#issuecomment-434280868 that behavior that comes from inferring a descriptor here, instead of deriving it directly from `desc` above, is a little unintuitive. IMO, it'd be nicer if `Expand` took an optional `std::vector<std::unique_ptr<Descriptor>>* output_descriptors` argument and filled it in parallel with the existing `output_scripts` argument. 

This is just a suggestion, though, and maybe something to consider for a future PR.

sipa replied:
That makes sense. The inferred descriptor is functionally equivalent with a more straightforward specialization of the descriptor you're suggesting, but descriptors are intended to at least be somewhat human readable, so this may matter.

However, it's quite a bit of code to implement that (it needs specialization code for a number of constructions, including `combo()` which would need to get split out into ``pkh()` and `pk()` etc), so I'd prefer to keep it for a later PR.

For that reason I also prefer not making any explicit promises about what is put in the RPC outputs; for now all it guarantees is something that encapsulates all the information. For `scantxoutset` we could indeed be better and maintain more of the input descriptor's structure.

ryanofsky reviewed (approved):
utACK 2286f206e4ea139477b88867a13a491d8d14db3d. Changes since last review are renaming descriptor fields to desc, adding them to scantxoutset output, and tweaking release notes.

sipa commented on test/functional/wallet_address_types.py:
```diff
@@ -99,6 +99,8 @@ def test_address(self, node, address, multisig, typ):
         """Run sanity checks on an address."""
         info = self.nodes[node].getaddressinfo(address)
         assert(self.nodes[node].validateaddress(address)['isvalid'])
+        assert('solvable' in info and info['solvable'])
```
Sorry I missed this, but I'd prefer not rewriting commits now.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -625,6 +625,80 @@ std::unique_ptr<Descriptor> ParseScript(Span<const char>& sp, ParseScriptContext
     return nullptr;
 }
 
+std::unique_ptr<PubkeyProvider> InferPubkey(const CPubKey& pubkey, ParseScriptContext, const SigningProvider& provider)
```
It can be used for extra sanity checking (like rejecting malformed constructions with an uncompressed pubkey inside a witness) early, instead of having them fail when expanding the inferred descriptor.

promag replied:
Could add that note to the code otherwise it can be removed in the future?

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -2140,7 +2142,11 @@ UniValue scantxoutset(const JSONRPCRequest& request)
                     throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Cannot derive script without private keys: '%s'", desc_str));
                 }
-                needles.insert(scripts.begin(), scripts.end());
+                for (const auto script : scripts) {
```
Could be `const auto&`?

promag reviewed (commented):
Commit adding release note could be reworded to "doc: Add release notes for #14477".

practicalswift commented on src/rpc/blockchain.cpp:
```diff
@@ -2140,7 +2142,11 @@ UniValue scantxoutset(const JSONRPCRequest& request)
                 }
-                needles.insert(scripts.begin(), scripts.end());
+                for (const auto script : scripts) {
+                    std::string desc = InferDescriptor(script, provider)->ToString();
```
Nit: Use another variable name to avoid showing `desc` declared on line 2133 (`auto desc = Parse(desc_str, provider);`)?

laanwj commented:
utACK 109699dd33c72539fc33619f7836b8088f63182c

ryanofsky reviewed (approved):
utACK 109699dd33c72539fc33619f7836b8088f63182c. Only superficial changes since last review: replacing assert with assert_equal, moving a variable declaration, avoiding copies with `&` and `std::move`.
